# Gene Expressions for different types of tumor
<div style="text-align: center;">
    <img src="images/gene_protein_cancer.jpg" alt="image" width="300" height="200">
</div>
[image src: https://www.cancer.gov/about-cancer/causes-prevention/genetics]

This project aims to identify different gene expressions associated to 5 types of tumor : 
- BRCA (Breast Cancer): Family of Genes (BRCA1 and BRCA2) are known as tumor suppresors. But mutation in these genes cause cancer.
- KIRC (Kidney Renal Clear Cell Carcinoma): 
- COAD (Colon Adenocarcinoma)
- LUAD (Lung Adenocarcinoma)
- PRAD (Prostate Adenocarcinoma)

The dataset is sourced from https://archive.ics.uci.edu/dataset/401/gene+expression+cancer+rna+seq.
The original dataset is published at https://www.synapse.org/Synapse:syn300013/discussion/threadId=5455. The Gene names in the dataset are dummy names. The actual gene names are at https://www.ncbi.nlm.nih.gov/gene, per this discussion thread https://www.synapse.org/Synapse:syn300013/discussion/threadId=5455. 

In [1]:
import warnings

warnings.filterwarnings("ignore")
DATA_ANALYSIS_DIR = "data-analysis/"

## Data Loading


In [2]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from numpy import ndarray
from pandas import DataFrame
from pandas import Series
from pandas.core.indexes.base import Index


from sklearn.model_selection import train_test_split

from src.utils.ArrayUtils import get_list_of_items_in_both_lists


In [3]:
# read the data from files
url_reconstructed = 'TCGA-PANCAN-HiSeq-801x20531/data.csv'

# url = 'https://drive.google.com/file/d/1VXyhDXpYT8G2Buhkc6kBjw93CLG1y1f0/view?usp=drive_link'
# # Use only the Id and reconstruct the URL
# url_reconstructed = 'https://drive.google.com/uc?id=' + url.split('/')[-2]

df = pd.read_csv(url_reconstructed)
tumor_df = pd.read_csv('TCGA-PANCAN-HiSeq-801x20531/labels.csv')
df

Unnamed: 0  gene_0    gene_1    gene_2    gene_3     gene_4  gene_5  \
0      sample_0     0.0  2.017209  3.265527  5.478487  10.431999     0.0   
1      sample_1     0.0  0.592732  1.588421  7.586157   9.623011     0.0   
2      sample_2     0.0  3.511759  4.327199  6.881787   9.870730     0.0   
3      sample_3     0.0  3.663618  4.507649  6.659068  10.196184     0.0   
4      sample_4     0.0  2.655741  2.821547  6.539454   9.738265     0.0   
..          ...     ...       ...       ...       ...        ...     ...   
796  sample_796     0.0  1.865642  2.718197  7.350099  10.006003     0.0   
797  sample_797     0.0  3.942955  4.453807  6.346597  10.056868     0.0   
798  sample_798     0.0  3.249582  3.707492  8.185901   9.504082     0.0   
799  sample_799     0.0  2.590339  2.787976  7.318624   9.987136     0.0   
800  sample_800     0.0  2.325242  3.805932  6.530246   9.560367     0.0   

       gene_6    gene_7  gene_8  ...  gene_20521  gene_20522  gene_20523  \
0    7.175175  0.591871     0.0  ...    4.926711    8.210257    9.723516   
1    6.816049  0.000000     0.0  ...    4.593372    7.323865    9.740931   
2    6.972130  0.452595     0.0  ...    5.125213    8.127123   10.908640   
3    7.843375  0.434882     0.0  ...    6.076566    8.792959   10.141520   
4    6.566967  0.360982     0.0  ...    5.996032    8.891425   10.373790   
..        ...       ...     ...  ...         ...         ...         ...   
796  6.764792  0.496922     0.0  ...    6.088133    9.118313   10.004852   
797  7.320331  0.000000     0.0  ...    6.371876    9.623335    9.823921   
798  7.536589  1.811101     0.0  ...    5.719386    8.610704   10.485517   
799  9.213464  0.000000     0.0  ...    5.785237    8.605387   11.004677   
800  7.957027  0.000000     0.0  ...    6.403075    8.594354   10.243079   

     gene_20524  gene_20525  gene_20526  gene_20527  gene_20528  gene_20529  \
0      7.220030    9.119813   12.003135    9.650743    8.921326    5.286759   
1      6.256586    8.381612   12.674552   10.517059    9.397854    2.094168   
2      5.401607    9.911597    9.045255    9.788359   10.090470    1.683023   
3      8.942805    9.601208   11.392682    9.694814    9.684365    3.292001   
4      7.181162    9.846910   11.922439    9.217749    9.461191    5.110372   
..          ...         ...         ...         ...         ...         ...   
796    4.484415    9.614701   12.031267    9.813063   10.092770    8.819269   
797    6.555327    9.064002   11.633422   10.317266    8.745983    9.659081   
798    3.589763    9.350636   12.180944   10.681194    9.466711    4.677458   
799    4.745888    9.626383   11.198279   10.335513   10.400581    5.718751   
800    9.139459   10.102934   11.641081   10.607358    9.844794    4.550716   

     gene_20530  
0      0.000000  
1      0.000000  
2      0.000000  
3      0.000000  
4      0.000000  
..          ...  
796    0.000000  
797    0.000000  
798    0.586693  
799    0.000000  
800    0.000000  

[801 rows x 20532 columns]

In [4]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 801 entries, 0 to 800
Columns: 20532 entries, Unnamed: 0 to gene_20530
dtypes: float64(20531), object(1)
memory usage: 125.5+ MB


In [5]:
print(f"Dataframe before adding the class column: {df.info()}")
print(f"Total # of columns: {len(df.columns)}")
print("COlumns type", type(df.columns))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 801 entries, 0 to 800
Columns: 20532 entries, Unnamed: 0 to gene_20530
dtypes: float64(20531), object(1)
memory usage: 125.5+ MB
Dataframe before adding the class column: None
Total # of columns: 20532
COlumns type <class 'pandas.core.indexes.base.Index'>


In [6]:
# Copy over the Class column into the main dataframe
df = df.rename(columns={'Unnamed: 0': 'sample'})
tumor_df = tumor_df.rename(columns={'Unnamed: 0': 'sample'})
df['Class'] = np.where( (df['sample'] == tumor_df['sample']), tumor_df['Class'], df['sample'])

# df['Class'] = tumor_df['Class']
# print(f"Total # of columns: {len(df.columns)}")

df.drop(columns=['sample'], axis=1, inplace=True)

df.head()


gene_0    gene_1    gene_2    gene_3     gene_4  gene_5    gene_6  \
0     0.0  2.017209  3.265527  5.478487  10.431999     0.0  7.175175   
1     0.0  0.592732  1.588421  7.586157   9.623011     0.0  6.816049   
2     0.0  3.511759  4.327199  6.881787   9.870730     0.0  6.972130   
3     0.0  3.663618  4.507649  6.659068  10.196184     0.0  7.843375   
4     0.0  2.655741  2.821547  6.539454   9.738265     0.0  6.566967   

     gene_7  gene_8  gene_9  ...  gene_20522  gene_20523  gene_20524  \
0  0.591871     0.0     0.0  ...    8.210257    9.723516    7.220030   
1  0.000000     0.0     0.0  ...    7.323865    9.740931    6.256586   
2  0.452595     0.0     0.0  ...    8.127123   10.908640    5.401607   
3  0.434882     0.0     0.0  ...    8.792959   10.141520    8.942805   
4  0.360982     0.0     0.0  ...    8.891425   10.373790    7.181162   

   gene_20525  gene_20526  gene_20527  gene_20528  gene_20529  gene_20530  \
0    9.119813   12.003135    9.650743    8.921326    5.286759         0.0   
1    8.381612   12.674552   10.517059    9.397854    2.094168         0.0   
2    9.911597    9.045255    9.788359   10.090470    1.683023         0.0   
3    9.601208   11.392682    9.694814    9.684365    3.292001         0.0   
4    9.846910   11.922439    9.217749    9.461191    5.110372         0.0   

   Class  
0   PRAD  
1   LUAD  
2   PRAD  
3   PRAD  
4   BRCA  

[5 rows x 20532 columns]

## Exploratory Data Analysis (EDA)
### Data Cleaning and PreProcessing

Data : 
- No NANs as stated on the data source page


In [7]:
df.isnull().any().any()

np.False_

In [8]:
# Find columns with 0 values
zero_cols = df.columns[(df == 0).all()]
print(f"# of columns with all 0s in them : {zero_cols}")
df = df.drop(columns=zero_cols, axis=1)
print(f"# of columns with all 0s in them : {df.columns[(df == 0).all()]}")

# of columns with all 0s in them : Index(['gene_5', 'gene_23', 'gene_4370', 'gene_4808', 'gene_4809', 'gene_4814',
       'gene_4816', 'gene_4817', 'gene_4831', 'gene_5288',
       ...
       'gene_18908', 'gene_18909', 'gene_18910', 'gene_18911', 'gene_18914',
       'gene_18915', 'gene_19450', 'gene_19451', 'gene_19452', 'gene_19671'],
      dtype='object', length=267)
# of columns with all 0s in them : Index([], dtype='object')


In [9]:
# Post cleanup, write the data to file
df.to_csv(DATA_ANALYSIS_DIR + "df_PostColumnCleanup.csv")


In [10]:
# Class analysis
class_unique_vals = df['Class'].unique()
assert(len(class_unique_vals) == 5)

# print the distribution
print(f"Distribution of Class values: \n{df['Class'].value_counts()}")


Distribution of Class values: 
Class
BRCA    300
KIRC    146
LUAD    141
PRAD    136
COAD     78
Name: count, dtype: int64


In [11]:
# Split -
X = df.drop(columns=['Class'], axis=1)
y = df['Class']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=df['Class'])


#### Summary : Data Cleaning and PreProcessing : 
- 0 NaNs
- 267 columns with only 0s as values. Dropped all of these columns
- 5 target values aka classifications


### Univariate and Multivariate Analysis
Given the large # of variables (columns), charting each out against other can be a challenge. Hence printing key insights such as stats to file.

In [12]:
from scipy import stats

df_desc = df.describe()
df_desc_t = df_desc.T
df_desc_t.to_csv(DATA_ANALYSIS_DIR + "df_describe.csv")

df_desc_zscore = df_desc.apply(stats.zscore)
df_desc_zscore.T.to_csv(DATA_ANALYSIS_DIR + "rawdata_zscore.csv")
# df_desc['upperbound'] = df_desc['mean'] + 3*(df_desc['std'])
# df_desc['lowerbound'] = df_desc['mean'] - 3*(df_desc['std'])


###### Identify Correlation across features / columns

In [ ]:
# --- NOTE : This take quite some time to run 11+ mins
# corr_df = X.corr()
# corr_df.to_csv(DATA_ANALYSIS_DIR + 'features_corr_matrix.csv')


In [ ]:
# Are there any Features with Correlation value= NaN ?
# -- In total there are 267 cols with 0 values, since they are deleted, result is zero
# corr_nan_list = corr_df.columns[corr_df.isna().all()].tolist()
# print(corr_nan_list)
# corr_df.head()

[]


gene_0    gene_1    gene_2    gene_3    gene_4    gene_6    gene_7  \
gene_0  1.000000  0.024111  0.027117 -0.080269 -0.028969  0.051855 -0.053647   
gene_1  0.024111  1.000000  0.533902  0.148348 -0.158555  0.202855  0.177154   
gene_2  0.027117  0.533902  1.000000  0.075988 -0.125376  0.334900  0.136246   
gene_3 -0.080269  0.148348  0.075988  1.000000  0.153666 -0.007905 -0.084841   
gene_4 -0.028969 -0.158555 -0.125376  0.153666  1.000000 -0.044347 -0.156698   

          gene_8    gene_9   gene_10  ...  gene_20521  gene_20522  gene_20523  \
gene_0 -0.024423 -0.012785 -0.015423  ...   -0.021109   -0.064341   -0.013874   
gene_1  0.023411  0.006425 -0.022650  ...    0.169172    0.190145    0.313725   
gene_2  0.013228 -0.045725  0.011893  ...   -0.021067    0.036411    0.242255   
gene_3 -0.033180  0.002434 -0.030119  ...   -0.186729   -0.014547   -0.011733   
gene_4  0.082124  0.098267 -0.042986  ...   -0.499208   -0.167764   -0.283480   

        gene_20524  gene_20525  gene_20526  gene_20527  gene_20528  \
gene_0   -0.008384   -0.036623    0.034314   -0.000368   -0.049142   
gene_1   -0.007284    0.123178   -0.255093    0.303771    0.247959   
gene_2    0.018755   -0.054584   -0.160446    0.163181    0.066709   
gene_3   -0.242855   -0.240551    0.092787   -0.000379    0.016733   
gene_4    0.179406   -0.367551    0.011739   -0.434604   -0.108036   

        gene_20529  gene_20530  
gene_0    0.041554   -0.027767  
gene_1   -0.067474    0.034348  
gene_2    0.011765    0.075604  
gene_3   -0.135829    0.023476  
gene_4   -0.007229    0.062752  

[5 rows x 20264 columns]

###### Output feature (column) characterstics based on the target variable

In [13]:
# Gene values (mean, mode, min, max, range) per class
# ---- NOTE : Below can take about 30secs to run.
# (X['gene_0'] == 0).sum()

def describeEachClass():
    for i in range(len(class_unique_vals)) :
        cls = class_unique_vals[i]
        # print('i= ', i, 'class= ', cls)
        df[df['Class'] == cls].describe().T.to_csv(f'{DATA_ANALYSIS_DIR}{cls}_describe_output.csv')

describeEachClass()
# print(type(df[df['Class'] == 'BRCA'].describe()))

##### Features with outliers

In [14]:
# Data has columns where values are close to 100% quartile. Get a list of these columns

def get_features_with_4thquartile_values_only(df) :
    '''
    @param df : Dataframe with all the data
    @return Dataframe consisting of descriptive stats for features in  in `df` where all values fall in 4th quartile.
    '''
    #generate descriptive statistics of given dataframe
    df_desc = df.describe()
    df_desc_t = df_desc.T

    print(f"Total # of columns: {len(df_desc_t)}")
    df_cols_outliers = df_desc_t[(df_desc_t['25%'] == 0) & (df_desc_t['50%'] == 0) & (df_desc_t['75%'] == 0)]
    print(f"# of columns with only outliers: {len(df_cols_outliers)}")

    return df_cols_outliers



In [15]:

# Identify if any of above columns have patterns such as associated with a type of tumor.
# Output the count of rows per Tumor type, for each of the above genes
def get_features_target_class_counts(df, df_cols_outliers, target_col='Class'):
    '''
    @param df : Dataframe with all the data
    @param df_cols_outliers : Dataframe with descriptive stats of features. See #get_features_with_4thquartile_values_only(df)
    @param target_col : Dependent variable in dataframe `df`.
    '''
    df_col_out_values = pd.DataFrame()
    for i in range(len(df_cols_outliers)):
        # for each record in outliers df, get only the rows that hv value greater than 0
        df_outliers_temp = df[df[df_cols_outliers.index[i]] > 0][[target_col, df_cols_outliers.index[i]]]
        df_outliers_temp1 = df_outliers_temp[target_col].value_counts().to_frame().T
        df_outliers_temp1['gene'] = df_cols_outliers.index[i]

        df_col_out_values = pd.concat([df_col_out_values, df_outliers_temp1], ignore_index=True)

        # df_col_out_values = pd.concat([df_col_out_values, df_outliers_temp], ignore_index=True)

    return df_col_out_values



##### Execution (of functions)

In [16]:
df_feature_outliers = get_features_with_4thquartile_values_only(df)
df_feature_outliers_count = get_features_target_class_counts(df, df_feature_outliers, 'Class')
df_feature_outliers_count.to_csv(f'{DATA_ANALYSIS_DIR}Outliers_cols_values_count.csv')


Total # of columns: 20264
# of columns with only outliers: 1660


In [17]:
import plotly.graph_objects as go
from plotly.subplots import make_subplots
import math

# init vars to help draw the layout
max_rows = 8
max_cols = 4
layout_diviser = min(max_rows, max_cols)

# vars to help place graph in the matrix
print(len(df_feature_outliers_count))
# no of features to plot per graph determined based on graph matrix size
features_per_graph = math.ceil( len(df_feature_outliers_count) / (max_rows * max_cols) )
start_index: int = 0
end_index: int = features_per_graph

# Total # of graphs to plot
total_graphs = math.ceil(len(df_feature_outliers_count) / features_per_graph)
print(f"# of feature groups: {total_graphs}")

# init plot
fig = make_subplots(rows=max_rows, cols=max_cols, shared_xaxes=True, vertical_spacing=0.03)

col_headers = df_feature_outliers_count.columns.to_list()

# Loop through # of graphs to plot. For each graph, plot the determined # of features
for fg in range(total_graphs):
    row = (fg) // layout_diviser
    col = (fg) % layout_diviser
    # print(f"row= {row} col= {col}")
    plot_df = df_feature_outliers_count.iloc[start_index:end_index, :]
    for i in range(len(plot_df)):
        fig.add_trace(go.Scatter(x=col_headers,
                                 y=plot_df.iloc[i, :-1],
                                 mode='lines',
                                 name=plot_df.loc[plot_df.index[i], 'gene']),
                                 row=row+1, col=col+1) #add_trace row and col, start index is 1, while row and col start with 0

    start_index = start_index + features_per_graph
    end_index = end_index + features_per_graph


fig.update_layout(height=1200, width=800, title_text="Outlier columns only : Records with (non-zero) values across the 5 classes")
fig.show()


1660
# of feature groups: 32


#### Summary : Univariate and Multivariate Analysis

- Descriptive Statistics of data : 1660 columns have mostly 0s, as values and their values fall into the 4th quartile.
- Outliers : Given the above 1660 columns / features have largely outliders, from above graph we identify the count of actual rows having these outlier values is less than 150, mostly less than 100. 
- Correlation : Given the large number of features, finding correlation between the features is challenging. The output file does not even open. Thus no inference to make.


### Feature Selection


### Dimensionality Reduction

#### PCA
To meet PCA needs, we need to ensure data confirms to below : 
- <u>correlation</u> between Independent vars.
- Variables must be <u>continuous</u> i.e numeric
- Variables must <u>be on same scale</u>. If one is in 1000s and another single digit, we will need to scale it.
- Data must be free from outliers

##### Remove Outliers
Based on data analysis on [outliers](#features-with-outliers) we know there are 1660 features with values in the 4th quartile. These gene expressions (features) are not specific to a tumor type (Class aka dependent variable) as observed in the output of function #get_features_target_class_counts. We will assume that these Gene expressions have not been recorded accurately or these have minimal influence from a tumor. Hence we remove these before running PCA.

In [18]:
# get quartiles for each column
def get_quartiles(col):
    quartile1, quartile3 = col.quantile([0.25, 0.75])
    inter_quartile_range = quartile3 - quartile1
    lower_range = quartile1 - (1.5* inter_quartile_range)
    upper_range = quartile3 + (1.5* inter_quartile_range)

    return lower_range, upper_range

# lowerQ, upperQ = get_quartiles(X_train['gene_1'])
# print(f'lower quantile= {lowerQ}, upper quantile= {upperQ}')

In [19]:
# cols = X_train_pca.columns

def normalize_within_quartiles(df_X:DataFrame, cols:Index):
    '''
    @input df_X : Dataframe with ONLY independent variables.
    '''
    for i in cols :
        LL, UL = get_quartiles(df_X[i])
        # if (i == 'gene_0'):
        #     print(f"UL: {np.where(df_X[i] > UL, UL, df_X[i])}")
            # print(f"{np.where(df_X[i] < LL, LL, df_X[i]}")
        df_X[i] = np.where(df_X[i] > UL, UL, df_X[i])
        df_X[i] = np.where(df_X[i] < LL, LL, df_X[i])
    return df_X



In [20]:
# print(type(df_feature_outliers.index.to_list()))
def clean_input_data(X_dataframe: DataFrame, df_feature_outliers: DataFrame):
    # print(len(X_train.columns))
    X_train_pca = X_dataframe.drop(columns=df_feature_outliers.index.to_list())
    # X_train_pca.columns
    X_train_pca = normalize_within_quartiles(X_train_pca, X_train_pca.columns)
    X_train_pca.head(10)
    print(f"Total # of cols after removing quartiles: {len(X_train_pca.columns)}")

    return X_train_pca

##### Prep for PCA

In [21]:
# PCA -
# -- Is PCA the right approach to this data set where we are trying to identify Gene expressions for tumors.
# -- Shudnt we include every expression - small or big ?
from sklearn.decomposition import PCA

pca = PCA(random_state=42)


In [238]:
def get_pca_features_weights(pca_model: PCA, classifier_model, feature_names: Index) :
    '''
    Identify the overall weights of each feature in terms of its contribution to the given model. From the given Classifier model
    it retrieves feature importances and this data is merged with the PCA components matrix.
    @param pca_model : PCA object, after it has been fit / trained on the data
    @param classifier_model : Model used for classification
    @param feature_names : List of all features used, before PCA was run.
    @return Series consisting of contribution of each feature to the model
    '''

    # get feature contributions for each Principal Component
    # feature_names = X_pre_pca_df.columns
    n_components = len(pca_model.components_)
    pca_components_df = pd.DataFrame(pca_model.components_.T,
                                     columns=[f'PC{i+1}' for i in range(n_components)],
                                     index=feature_names)


    # Random forest importances for each component
    if (hasattr(classifier_model, 'feature_importances_')):
        rf_importances = pd.Series(
            classifier_model.feature_importances_,
            index=[f'PC{i+1}' for i in range(n_components)]
        )

        # --- calculate original feature importance by weighted combination ---
        orig_importances = pca_components_df.dot(rf_importances).abs() # we only care abt the magnitude
        # sum of all shud be 1, hence find each value's contrib to 100%
        orig_importances = orig_importances / orig_importances.sum()
        return orig_importances.sort_values(ascending=False)

    return None



##### Execution (Feature Selection)

In [24]:
X_train_pca = clean_input_data(X_train, df_feature_outliers)
X_train_pca.head(10)

Total # of cols after removing quartiles: 18604


gene_1    gene_2    gene_3     gene_4    gene_6    gene_7   gene_10  \
391  4.094565  4.704501  7.545559   9.752299  8.595821  0.859970  1.152183   
148  6.237034  5.043235  6.297397  10.391415  7.669941  0.913033  0.527571   
595  3.028304  2.156138  6.459537   9.488378  5.319195  0.000000  1.043764   
770  2.706995  3.574586  6.090422   9.557619  7.162452  0.467175  0.467175   
493  3.522621  3.463021  8.133024  10.431341  7.536667  0.000000  0.000000   
706  3.913167  4.064193  6.530920   9.702678  8.404367  0.566084  1.547055   
526  2.463492  3.442293  6.850874   9.415991  7.270968  0.000000  0.485015   
480  2.676809  3.955629  6.639035  10.301165  7.946497  0.808426  0.460166   
223  2.870404  1.268614  6.516157   9.287569  6.674658  0.825541  0.000000   
455  3.069737  3.623200  6.744955   9.591219  7.254254  0.000000  2.204846   

      gene_11   gene_12   gene_13  ...  gene_20520  gene_20521  gene_20522  \
391  2.222650  2.222650  0.000000  ...    9.532568    4.598568    7.654557   
148  0.527571  2.752642  0.000000  ...    8.136238    5.669299    8.437523   
595  0.000000  3.703887  0.000000  ...    7.647998    6.436851    8.557640   
770  2.151112  2.667597  0.467175  ...    9.467758    6.261474    9.005094   
493  1.497740  1.901494  0.000000  ...   10.234063    4.848788    8.581671   
706  2.652119  1.547055  0.971663  ...    9.741261    5.301701    8.260317   
526  0.000000  3.458119  0.000000  ...    7.722398    6.622088    9.022593   
480  0.460166  3.933100  0.000000  ...    7.855828    6.306275    9.071972   
223  0.471031  1.109895  0.000000  ...    8.173152    7.017599    8.560134   
455  1.765450  3.268928  0.000000  ...    7.722623    6.449447    8.276566   

     gene_20523  gene_20524  gene_20525  gene_20526  gene_20527  gene_20528  \
391   10.089887    7.107677    8.832156   12.816038    9.317229    9.461801   
148   10.524503    1.467801    9.666187   11.258572   10.062033    9.018030   
595    9.749657    5.406037   10.590297   11.449572    9.303614    9.228160   
770    9.897212    6.405101    9.971630   12.057508    9.898369    9.835752   
493    9.682631    0.542010    8.867365   12.051070   10.181786    8.857471   
706    9.720046    5.513810    9.539921   12.177852    9.465385    9.314588   
526   10.219907    2.886277   10.212630   11.027982   10.388082    9.968667   
480   10.136145    7.163640   10.286477   10.573922    9.523879   10.355120   
223    9.703589    5.989782   10.192428   11.646392   10.495325    9.885373   
455    9.765221    3.339038   10.245957   12.200892   10.292529    9.378830   

     gene_20529  
391    9.258929  
148    7.737159  
595    4.403943  
770    4.649144  
493    2.792439  
706    3.946366  
526    7.738680  
480    3.525868  
223    4.199633  
455    5.152070  

[10 rows x 18604 columns]

In [25]:
# pca_df.to_csv(f"{DATA_ANALYSIS_DIR}PCA_ExpVarianceRatio.csv")
# pca_df[0:640]
# #TODO : How to get column names as part of PCA output. Also cleanup rows beyond 639

#### Summary : Feature Selection
- Features with outliers values (in 4th quartile) dropped
- PCA reduces features to 640, from 18604 features. Data prepared for PCA execution in next section, as part of the Pipeline.


## Modeling

##### Utility Functions for Graphing 

In [28]:
# Function to plot Confusion Matrix
def plot_confusion_matrix(conf_matrix:ndarray, labels:ndarray, title="Confusion Matrix"):

    conf_matrix = conf_matrix[::-1]
    # labels = labels[::-1]

    #heat map
    fig = go.Figure(data=go.Heatmap(
        z=conf_matrix,
        x=labels,
        y=labels[::-1], #reverse the order to align labels with way Conf matrix is output
        colorscale='Rainbow', # 'Hot', # 'YlOrRd', # 'YlGnBu', #'Viridis',
        texttemplate="%{z}",
        textfont={"size": 10}
    ))

    fig.update_layout(
        title_text = title,
        xaxis_title="Predicted Class",
        yaxis_title="Actual Class",
        # xaxis={'side': 'top'},
        # yaxis={'autorange': 'reversed'},
        width=500,
        height=500,
    )

    fig.show()



In [142]:

# Using plotly's graph_objects
def plot_feature_importance_comparison_plotly(pca_model: PCA, classifier_model, feature_importances: Series, X_pre_pca_df: DataFrame):

    # top_features = feature_importances.head(10)

    # fig = px.bar(features_df, x=features_df.index, y=features_df[0])
    fig = go.Figure()
    fig.add_trace(go.Bar(
        x=feature_importances.index,
        y=feature_importances.values,
        marker=dict(color='indianred'),
        marker_color='indianred'
    ))
    fig.update_layout(title=f'Gene expression contribution to the model\n{classifier_model}', template='plotly_white')
    fig.show()


## Using Matplotlib
def plot_feature_importance_comparison(pca_model: PCA, classifier_model, feature_importances: DataFrame, X_pre_pca_df: DataFrame):

    feature_names = X_pre_pca_df.columns


    fig, axes = plt.subplots(1, 3, figsize=(15,6))

    #1. Original feature importances
    top_features = feature_importances.head(10)
    axes[0].barh(range(len(top_features)), top_features.values)
    axes[0].set_yticks(range(len(top_features)))
    axes[0].set_yticklabels(top_features.index)
    axes[0].set_xlabel('Importance')
    axes[0].set_title('Original feature importance\n(PCA + RF Combined)')
    axes[0].invert_yaxis()

    #2 PCA importances from Random Forest
    rf_importances = pd.Series(
        classifier_model.feature_importances_,
        index=[f'PC{i+1}' for i in range(classifier_model.n_features_in_)]
    )
    axes[1].bar(range(len(rf_importances)), rf_importances.values)
    axes[1].set_xlabel('Principal Component')
    axes[1].set_ylabel('RF Importance')
    axes[1].set_title('PC Importance\n(Random Forest)')
    axes[1].set_xticks(range(len(rf_importances)))
    axes[1].set_xticklabels([f'PC{i+1}' for i in range(len(rf_importances))], rotation=45)

    # 3. Explained variance ratio of PCs
    axes[2].bar(range(len(pca_model.explained_variance_ratio_)), pca_model.explained_variance_ratio_)
    axes[2].set_xlabel('Principal Component')
    axes[2].set_ylabel('Explained Variance Ratio')
    axes[2].set_title('PC Explained Variance\n(PCA)')
    axes[2].set_xticks(range(len(pca_model.explained_variance_ratio_)))
    axes[2].set_xticklabels([f'PC{i+1}' for i in range(len(pca_model.explained_variance_ratio_))], rotation=45)

    plt.tight_layout()
    plt.show()



In [243]:
def analyze_tree_path(pca_model: PCA, classifier_model, X_pca_df: DataFrame, y_series:Series, feature_names: Index, class_names=class_unique_vals):
    '''
    For each class / target, analyze the contribution of each feature by traversing the tree, finding the leaf node and
    then merging that with each feature's individual weights, calculated using the function #get_pca_features_weights.
    @param pca_model : PCA model, after the data has been fit aka trained on the data.
    @param classifier_model : Model used for classification, after it has been trained on the data
    @param X_pca_df : The DataFrame after PCA has been run and data has been transformed.
    @param y_series : The Y series corresponding to above X_pca_df dataset i.e if above is `test` dataset, this should also be test dataset
    @param feature_names : All the features used prior to running PCA
    @param class_names : List of unique Classes / Target Variables, the data represents. In this case we have the 5 tumors.
    @return DataFrame with columns as the 4 target classes and rows as features aka gene expressions.
    '''
    # feature_names = X_pca_df.columns
    #init dictionary that will hold each class details
    class_importances = {class_name: np.zeros(len(feature_names)) for class_name in class_unique_vals}

    #loop through each DecisionTree used by the model
    for tree in classifier_model.estimators_:
        tree_importances = tree.feature_importances_
        # get index of leaf node where sample is predicted
        leaf_nodes = tree.apply(X_pca_df)

        for row_idx, leaf_id in enumerate(leaf_nodes):
            # leaf_nodes has only the node number. Not the index that matches against corresponding index in y_series. Hence we use X_pca_df to get the index
            #row_idx is sequential increment of leaf_nodes
            record_index = X_pca_df.index[row_idx]
            predicted_class = y_series[record_index]

            original_importances = get_pca_features_weights(pca_model, classifier_model, feature_names)
            class_importances[predicted_class] += original_importances


    #Normalize
    for class_name in class_names:
        total = np.sum(class_importances[class_name])
        if total > 0:
            class_importances[class_name] /= total

    return pd.DataFrame(class_importances)



#### Random Forest

In [26]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import plot_tree
from sklearn.metrics import roc_curve
from sklearn.metrics import roc_auc_score
from sklearn.metrics import auc
from sklearn.preprocessing import label_binarize
from sklearn.preprocessing import LabelBinarizer
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score
from sklearn.metrics import classification_report
import shap


In [27]:
def createPipeline(model):

    pca_pipeline = Pipeline([
        ('scaler', StandardScaler()),
        ('pca', pca),
        ('model', model)
    ])
    return pca_pipeline


In [29]:
# --- Create the model and fit train data
def fit_randomforest(X_train_pca: DataFrame, y_train: Series):
    rd_classifier = RandomForestClassifier(random_state=42, oob_score=True)
    rd_pipeline = createPipeline(rd_classifier)
    rd_pipeline.fit(X_train_pca, y_train)

    return rd_pipeline



In [245]:
# evaluate the model
def predict_randomforest(rd_pipeline: Pipeline, X_pca_dataframe: DataFrame, y_series: Series) :
    '''
    @param rd_pipeline : sklearn Pipeline object with RandomForest as a classifier model
    @param X_pca_dataframe : DataFrame of independent vars prior to running PCA
    @param y_series : Series object consisting of dependent data associataed to above X dataset.
    @return Returns the below in the given order :
    predict_vals : Values predicted by the model, an ndarray
    rd_train_accuracy : Value returned by the `.accuracy_score` method
    rd_train_proba_accuracy : Value returned by `accuracy_score` using probability prediction
    report : Classification report with details of precision, recall, f1score, # of records tied to each class etc.. Refer to sklearn.metrics.classification_report
    '''

    rf_predict = rd_pipeline.predict(X_pca_dataframe)

    # The .score method uses accuracy_score, underneath.
    # rd_train_score = rd_pipeline.score(X_pca_dataframe, y_series)
    # print(f"Random Classifier Model Score= {rd_train_score}")

    # using actual predictions
    rd_train_accuracy = accuracy_score(y_series, rf_predict)

    # converting probabilities to predictions
    # below outputs an array for each row with proba of a match [[0.79, 0.05, 0.04, 0.09, 0.03]
    rd_train_pred_proba = rd_pipeline.predict_proba(X_pca_dataframe)
    # below converts abv array to index with max prob value. eg: for above example it outputa "0" as the index given value at this index has max prob
    rd_train_pred_indices = np.argmax(rd_train_pred_proba, axis=1)
    rd_train_pred_accuracy = accuracy_score(y_series, rd_train_pred_indices)

    # Print the whole Classification report
    report = classification_report(y_series, rf_predict)

    # Using SHAP to explain - #TODO
    # explainer = shap.Explainer(rd_model)
    # shap_values = explainer(X_pca_dataframe)
    # original_shap = shap_values.values @ pca_model.components_

    return rf_predict, rd_train_accuracy, rd_train_pred_accuracy, report

# print(f"rd_oob_score = {rd_model.oob_score_}")
# print(f"PCA No. of components= {pca_model.n_components_}")
# print(f"PCA Explained Variance Ratio= {pca_model.explained_variance_ratio_}")
# print(f"rd_feature_imp = {rd_model.feature_importances_}")


##### Model Evaluation

In [244]:
from sklearn.metrics import confusion_matrix

def run_pca_randomforest():
    # Fit and run for train data
    print('----------- Train data -------------')
    rf_pca_pipeline = fit_randomforest(X_train_pca, y_train)
    # rf_predict_train, rf_train_accuracy, rf_train_proba_accuracy = predict_randomforest(rf_pca_pipeline, X_train_pca, y_train)

    # Test data
    print('----------- Test data -------------')
    rf_X_test_pca = clean_input_data(X_test, df_feature_outliers)
    rf_predict_test, rf_test_accuracy, rf_test_proba_accuracy, cls_report = predict_randomforest(rf_pca_pipeline, rf_X_test_pca, y_test)
    print(f"Random Classifier Accuracy score= {rf_test_accuracy}")
    print(f"Accuracy score after converting prob to predictions= {rf_test_proba_accuracy}")
    print(f"Printing the Classification Report: \n",cls_report)

    # confusion matrix
    cm = confusion_matrix(y_true=y_test, y_pred=rf_predict_test, labels=class_unique_vals)
    plot_confusion_matrix(cm, labels=class_unique_vals, title='Random Forest with PCA : Confusion Matrix')


    # --- Deeper dive into understanding contributions of each feature ----
    pca_model = rf_pca_pipeline.named_steps['pca']
    rf_model = rf_pca_pipeline.named_steps['model']

    # PCA : identify features with highest weightage :
    pca_weights_df = get_pca_features_weights(pca_model=pca_model,
                                              classifier_model=rf_model,
                                              feature_names=rf_X_test_pca.columns)
    if pca_weights_df is not None:
        # plot_feature_importance_comparison(pca_model, rd_model, pca_weights_df, X_pca_dataframe)
        plot_feature_importance_comparison_plotly(pca_model, rf_model, pca_weights_df, rf_X_test_pca)
    else :
        print(f'PCA Weights for {rf_model} could not be calculated')


    transformed_data = pca_model.transform(rf_X_test_pca)
    transformed_df = pd.DataFrame(transformed_data, columns=pca_model.get_feature_names_out(), index=rf_X_test_pca.index)
    gene_contrib_2_tumor = analyze_tree_path(pca_model,
                            classifier_model=rf_model,
                            X_pca_df=transformed_df,
                            feature_names=rf_X_test_pca.columns,
                            y_series=y_test)
    print(f"Writing {len(gene_contrib_2_tumor)} Gene expression contributions for each Tumor to CSV file: ", )
    gene_contrib_2_tumor.to_csv(DATA_ANALYSIS_DIR + 'RForest_Gene_exp_contrib_2each_Tumor.csv')



In [246]:
run_pca_randomforest()

----------- Train data -------------
----------- Test data -------------
Total # of cols after removing quartiles: 18604
Random Classifier Accuracy score= 0.8819875776397516
Accuracy score after converting prob to predictions= 0.0
Printing the Classification Report: 
               precision    recall  f1-score   support

        BRCA       0.77      1.00      0.87        60
        COAD       1.00      0.69      0.81        16
        KIRC       1.00      0.93      0.97        30
        LUAD       1.00      0.64      0.78        28
        PRAD       0.96      0.93      0.94        27

    accuracy                           0.88       161
   macro avg       0.95      0.84      0.88       161
weighted avg       0.91      0.88      0.88       161



Writing 18604 Gene expression contributions for each Tumor to CSV file: 


In [ ]:
# Area under Curve - REF : https://www.geeksforgeeks.org/interpreting-random-forest-classification-results/
# -------- #TODO : need to fix the roc_curve function
def plot_aoc_randomforest(rd_test_pred_proba: ndarray, y_test: Series) :
    target_vals = y_test.unique()
    # y_test_bin = label_binarize(y_test, classes=[0,1,2,3,4])
    label_binzer = LabelBinarizer()
    label_binzer.fit(y_test)
    y_test_bin = np.array(label_binzer.transform(y_test))
    print(y_test_bin[0])
    fpr = dict()
    tpr = dict()
    roc_auc = dict()

    print(f"y_test_bin : {y_test_bin[1]}")
    print(f"rd_preds_prob : {rd_test_pred_proba[:, 1]}")
    # print(f"y_test_bin : {len(y_test_bin[:, 1])}")
    # print(f"rd_preds_prob : {rd_preds_prob[:, 1]}")
    # print(rd_preds_prob)

    for index in range(len(target_vals)):
        fpr[index], tpr[index], _ = roc_curve(y_test_bin[index], rd_test_pred_proba[:, index])
        # print(f"FPR at {index}: \n{fpr[index]}")
        # print(f"TPR at {index}: \n{tpr[index]}")
        roc_auc[index] = auc(fpr[index], tpr[index])

    # Plot ROC curve
    plt.figure()
    for index in range(len(target_vals)) :
        plt.plot(fpr[index], tpr[index], lw=2, label=f"ROC curve of class {target_vals[index]} (area = {roc_auc[index]:.2f})")

    # plt.plot([0,1], [0,1], color='navy', lw=2, linestyle='--')
    plt.xlim([0.0, 1.0])
    plt.ylim([0.0, 1.05])
    plt.xlabel('False Positive Rate')
    plt.ylabel('True Positive Rate')
    plt.title('Receiver Operating Characterstic for Tumor classes')
    plt.legend(loc="lower right")
    plt.show()



#### XGBoost Model

In [147]:
from xgboost import XGBClassifier

def fit_xgb(xgb_X_train: DataFrame, xgb_y_train: ndarray):
    xgb_classifier = XGBClassifier(use_label_encoder=False, eval_metric='mlogloss')
    xgb_pipeline = createPipeline(xgb_classifier)
    xgb_pipeline.fit(xgb_X_train, xgb_y_train)

    return xgb_pipeline

In [151]:
def predict_xgb(xgb_pipeline:Pipeline, X_pca_dataframe:DataFrame, y_series:ndarray) :
    '''
    @param : sklearn Pipeline object with XGB as a classifier model
    @return Returns the below in the given order :
    xgb_predict: Values predicted by the model, as ndarray
    xgb_train_score : Value returned by the `.score` method
    xgb_train_accuracy : Value returned by the `.accuracy_score` method
    xgb_train_proba_accuracy : Value returned by `accuracy_score` using probability prediction
    '''
    xgb_predict = xgb_pipeline.predict(X_pca_dataframe)

    xgb_pca_model = xgb_pipeline.named_steps['pca']
    xgb_model = xgb_pipeline.named_steps['model']

    # using actual predictions
    xgb_train_accuracy = accuracy_score(y_series, xgb_predict)
    print(f"XGBooster Classifier Accuracy score= {xgb_train_accuracy}")
    # converting probabilities to predictions
    # below outputs an array for each row with proba of a match [[0.79, 0.05, 0.04, 0.09, 0.03]
    xgb_train_pred_proba = xgb_pipeline.predict_proba(X_pca_dataframe)

    # below converts abv array to index with max prob value. eg: in case of above example output 0 as the index i.e max prob
    xgb_train_proba_indices = np.argmax(xgb_train_pred_proba, axis=1)
    y_series_indices = np.argmax(y_series, axis=1)
    xgb_train_proba_accuracy = accuracy_score(y_series_indices, xgb_train_proba_indices)
    print(f"Accuracy score after converting prob to predictions= {xgb_train_proba_accuracy}")

    # Print the whole Classification report
    report = classification_report(y_series, xgb_predict)
    print(report)

    # PCA : identify features with highest weightage :
    pca_weights_df = get_pca_features_weights(xgb_pca_model, xgb_model, X_pca_dataframe, class_unique_vals)
    if pca_weights_df is not None:
        # plot_feature_importance_comparison(pca_model, rd_model, pca_weights_df, X_pca_dataframe)
        plot_feature_importance_comparison_plotly(xgb_pca_model, xgb_model, pca_weights_df, X_pca_dataframe)
    else :
        print(f'PCA Weights for {xgb_model} could not be calculated')


    # print(f"PCA No. of components= {xgb_pca_model.n_components_}")

    return xgb_predict, xgb_train_accuracy, xgb_train_proba_accuracy


##### Model Evaluation (XGB)

In [154]:

def run_pca_xgboost():
    # convert Target to binary given XGBoost requires it
    print('------------ Train data ------------')
    label_binzer = LabelBinarizer()
    label_binzer.fit(y_train)
    y_train_bin = np.array(label_binzer.transform(y_train))
    # Uncomment below to identify mappings between binary format and the actual label
    # 0=BRCA 1=COAD, 2=KIRC, 3=LUAD, 4=PRAD
    # print("Binary data: ", y_train_bin)
    # print("Label to int mapping: ", label_binzer.inverse_transform(np.array(y_train_bin)))

    # Fit and run for train data
    xg_pca_pipeline = fit_xgb(X_train_pca, y_train_bin)
    # predict_train, xg_train_accuracy, xg_train_proba_accuracy = predict_xgb(xg_pca_pipeline, X_train_pca, y_train_bin)

    # Test data
    print('------------ Test data ------------')
    label_binzer.fit(y_test)
    y_test_bin = np.array(label_binzer.transform(y_test))
    xg_X_test_pca = clean_input_data(X_test, df_feature_outliers)
    predict_test, xg_test_accuracy, xg_test_proba_accuracy = predict_xgb(xg_pca_pipeline, xg_X_test_pca, y_test_bin)

    y_test_bin_arr = np.array(y_test_bin)
    # confusion matrix
    cm = confusion_matrix(y_true=np.argmax(y_test_bin_arr, axis=1), y_pred=np.argmax(predict_test, axis=1)) #, labels=class_unique_vals)
    plot_confusion_matrix(cm, labels=class_unique_vals, title='XGBoost with PCA : Confusion Matrix')


# xgb_pca_pipeline = fit_xgb(X_train_pca, y_train)
# xgb_train_score, xgb_train_accuracy, xgb_train_proba_accuracy = predict_xgb_train(xgb_pca_pipeline)
# # X_test_pca = clean_test_data(X_test)

# xgb_test_score, xgb_test_accuracy, xgb_test_proba_accuracy = predict_xgb_test(xgb_pca_pipeline, X_test_pca, y_test)



In [155]:
run_pca_xgboost()

------------ Train data ------------
------------ Test data ------------
Total # of cols after removing quartiles: 18604
XGBooster Classifier Accuracy score= 0.906832298136646
Accuracy score after converting prob to predictions= 0.9751552795031055
              precision    recall  f1-score   support

           0       0.98      1.00      0.99        60
           1       0.93      0.88      0.90        16
           2       1.00      0.97      0.98        30
           3       1.00      0.68      0.81        28
           4       1.00      0.96      0.98        27

   micro avg       0.99      0.92      0.95       161
   macro avg       0.98      0.90      0.93       161
weighted avg       0.99      0.92      0.95       161
 samples avg       0.91      0.92      0.92       161

Model has feature_importance_ attr ? True
